### PR TITLE
feat: add hasher to ProofOptions

### PR DESCRIPTION
**Pull request by vlopes11**
With the introduction of winterfell 0.5, the hasher is now an argument of proofs, and it will define the target security level as part of its hasher

This commits adapts air `ProofOptions` to have such attribute.

### CODE REVIEW

File: miden/src/cli/prove.rs [hunk 1]
```diff
@@ -61,9 +53,24 @@ impl ProveCmd {
         let advice_provider = input_data.parse_advice_provider()?;
 
         // execute program and generate proof
-        let (stack_outputs, proof) =
-            prover::prove(&program, stack_inputs, advice_provider, &self.get_proof_security())
-                .map_err(|err| format!("Failed to prove program - {:?}", err))?;
+        let proof = match self.security.as_str() {
+            "96bits" => prover::prove(
```
Comment (hackaugusto) [inline miden/src/cli/prove.rs:57]: maybe use [`FromStr`](https://doc.rust-lang.org/std/str/trait.FromStr.html)?

File: prover/src/lib.rs [hunk 2]
```diff
@@ -107,10 +118,14 @@ impl ExecutionProver {
...
 
-impl Prover for ExecutionProver {
+impl<H> Prover for ExecutionProver<H>
+where
+    H: Clone + Hasher,
+{
     type Air = ProcessorAir;
     type BaseField = Felt;
     type Trace = ExecutionTrace;
+    // TODO should be H
```
Comment (hackaugusto) [inline prover/src/lib.rs:128]: these TODOs should probably have its own issue if not done on this PR, no?

File: air/src/options.rs [hunk 3]
```diff
@@ -1,11 +1,28 @@
...
-impl ProofOptions {
+/// A set of arguments to specify how the Miden proofs should be constructed.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct ProofOptions<H>
+where
+    H: Hasher,
+{
+    hasher: PhantomData<H>,
+    options: WinterProofOptions,
+}
```
Comment (bobbinth) [inline air/src/options.rs:16]: I would do it differently:

I would introduce an enum for hash functions (e.g., `HashFunction`) we want to support. We probably don't need too many options. For now, we can have just 3: `Blake3_192`, `Blake3_256`, and `RPO256`.

Then, this struct could look something like:
```Rust
pub struct ProofOptions {
    options: WinterProofOptions,
    hash_fn: HashFunction,
}
```

File: prover/src/proof.rs [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+use core::ops::Deref;
+
+// MIDEN PROOF
+// ================================================================================================
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct MidenProof {
+    outputs: StackOutputs,
+    proof: StarkProof,
+}
```
Comment (bobbinth) [inline prover/src/proof.rs:11]: I'm not sure we need this struct - or at least not in this shape. I would not combine `outputs` and `proof` into a single object as they could come from different places.

File: prover/src/lib.rs [hunk 5]
```diff
@@ -33,14 +36,15 @@ pub use winter_prover::StarkProof;
...
 /// Returns an error if program execution or STARK proof generation fails for any reason.
-pub fn prove<A>(
+pub fn prove<A, H>(
     program: &Program,
     stack_inputs: StackInputs,
     advice_provider: A,
-    options: &ProofOptions,
-) -> Result<(StackOutputs, StarkProof), ExecutionError>
+    options: &ProofOptions<H>,
+) -> Result<MidenProof, ExecutionError>
```
Comment (bobbinth) [inline prover/src/lib.rs:44]: Let's not add `H` as the generic parameter here. Ideally, the info about which hash function to use should come from `ProofOptions` via an enum.

However, we may need to wrap `StarkProof` into something which also contains information about the hash function used.

File: prover/src/proof.rs (hunk 4 above)
Comment (vlopes11) [reply prover/src/proof.rs:11]: Done!

Comment (vlopes11) [reply prover/src/proof.rs:11]: Sounds good; the idea is to always avoid returning tuples as they can sometimes be confusing and cannot be documented. However, that isn't mandatory for this PR and was just an extra improvement

File: air/src/options.rs (hunk 3 above)
Comment (vlopes11) [reply air/src/options.rs]: Done!

File: prover/src/lib.rs (hunk 2 above)
Comment (vlopes11) [reply prover/src/lib.rs:118]: Agreed! I usually leave these questions to be checked @ review and create the issues after discussion/before merging

File: miden/src/cli/prove.rs (hunk 1 above)
Comment (vlopes11) [reply miden/src/cli/prove.rs:57]: That is something to think about. This is not really a property of the proof options, but instead it is a CLI argument that we accept. I wonder if we should couple the two

**Review by bobbinth — APPROVED** [review]
Looks good! Thank you!

**Pull request merged**